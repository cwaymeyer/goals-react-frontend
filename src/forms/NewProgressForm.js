import { useState, useContext } from "react";
import {
  Container,
  Box,
  Grid,
  Button,
  TextField,
  InputAdornment,
} from "@mui/material";
import ProgressContext from "../context/progressContext";
import calculate1RM from "../helpers/calculate1RM";

function NewProgressForm({ createProgress, goal_id }) {
  const { setProgressData } = useContext(ProgressContext);
  const initialState = {
    goal_id: goal_id,
    weight: 0,
    reps: 0,
    date: "",
  };

  const [formData, setFormData] = useState(initialState);

  function handleChange(e) {
    const { name, value } = e.target;
    setFormData((formData) => ({ ...formData, [name]: value }));
  }

  async function handleSubmit(e) {
    e.preventDefault();
    formData.orm = +calculate1RM(formData.weight, formData.reps);
    formData.date = Date.now();
    setProgressData((progressData) => [
      ...progressData,
      { x: formData.date, y: formData.orm },
    ]);
    await createProgress(formData);
    setFormData(initialState);
  }

  return (
    <Container maxWidth="sm">
      <Box sx={{ m: 3, mt: 6, mb: -4 }}>
        <h3>Record workout 🏋️‍♀️</h3>
      </Box>
      <form onSubmit={handleSubmit}>
        <Grid container p={3}>
          <Grid item xs={6} my={1} pr={1}>
            <TextField
              label="Weight"
              name="weight"
              type="number"
              value={formData.weight}
              onChange={handleChange}
              sx={{ width: 1 / 1 }}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">lb</InputAdornment>
                ),
              }}
            />
          </Grid>
          <Grid item xs={6} my={1} pl={1}>
            <TextField
              label="Reps"
              name="reps"
              type="number"
              value={formData.reps}
              onChange={handleChange}
              sx={{ width: 1 / 1 }}
            />
          </Grid>
          <Button
            type="submit"
            variant="contained"
            color="primary"
            onSubmit={handleSubmit}
            sx={{ mt: 2 }}
          >
            Submit
          </Button>
        </Grid>
      </form>
    </Container>
  );
}

export default NewProgressForm;
