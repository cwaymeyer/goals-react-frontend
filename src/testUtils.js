import UserContext from "./context/userContext";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

const progress = [
  {
    id: 10,
    goal_id: 1,
    weight: 50,
    reps: 5,
    orm: "56.3",
    date: "1645477469350",
  },
  {
    id: 11,
    goal_id: 1,
    weight: 50,
    reps: 6,
    orm: "58.1",
    date: "1645477669350",
  },
];

const starting_progress = {
  orm: "56.3",
  date: "1645477269350",
};

const latest_progress = {
  orm: "58.1",
  date: "1645477669350",
};

const goal = {
  id: "1",
  name: "Test Goal",
  username: "demoUser",
  start_weight: null,
  target_weight: 100,
  timeline: 3,
  start_date: "1645477269350",
  end_date: "1653163269350",
  progress: progress,
  starting_progress: starting_progress,
  latest_progress: latest_progress,
};

const demoUser = {
  username: "demoUser",
  password: "password",
  goals: [goal],
};

const UserProvider = ({ children, currentUser = demoUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

const renderWithRouter = (ui, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);

  return render(ui, { wrapper: BrowserRouter });
};

export {
  UserProvider,
  renderWithRouter,
  goal,
  starting_progress,
  latest_progress,
};
