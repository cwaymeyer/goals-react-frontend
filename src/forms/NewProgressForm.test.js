import { render } from "@testing-library/react";
import { UserProvider, latest_progress } from "../testUtils";
import NewProgressForm from "./NewProgressForm";

it("renders without crashing", function () {
  render(
    <UserProvider>
      <NewProgressForm latestProgress={latest_progress} />
    </UserProvider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <UserProvider>
      <NewProgressForm latestProgress={latest_progress} />
    </UserProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
