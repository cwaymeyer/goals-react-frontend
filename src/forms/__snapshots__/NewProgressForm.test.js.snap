// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
    xs="[object Object]"
  >
    <h3>
      Record workout
    </h3>
    <form>
      <div
        class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-ojs117-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="mui-5"
              id="mui-5-label"
            >
              Weight
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-5"
                name="weight"
                type="number"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                >
                  lb
                </p>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Weight
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-2sjwqm-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="mui-6"
              id="mui-6-label"
            >
              Reps
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-6"
                name="reps"
                type="number"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Reps
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-13lm9dl-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="mui-7"
              id="mui-7-label"
            >
              Date
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                aria-label="Choose date"
                aria-readonly="true"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-7"
                readonly=""
                type="text"
                value="02/26/2022"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1z7n62"
                >
                  <span>
                    Date
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-fxvftd-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Submit
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9 css-zngqn1-MuiGrid-root"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
